examples/machine_learning/src/lib.rs

```diff
@@ -138,63 +138,111 @@ fn evaluate_prediction(hits: &mut u32, animal: &Animal, prediction: &[f64]) -> (
     (actual_type, accurate)
 }
 
-struct Node;
+struct Node {
+    training_set_size: u32,
+    test_set_size: u32,
+    training_matrix: Option<Matrix<f64>>,
```

Comment:
Instead of a struct of Options that are always either all set or all unset, I suggest factoring the optionality out and make this something like `Option<Config>` where `Config` is a struct with the various matrices you need. This way you only need to pattern match once on the config and you get all your values out of it, rather than pattern match on individual values and still having to deal with the (impossible) case of some of these being set and some being unset. In fact you may as well move even the other non-matrix parameter inside the config struct, since there is little or no use of them unless all of them are actually present.